File: swiftPrep/Sources/Sorting/InsertionSort.swift

/**
	Sort the sequence in ascending order in place.

	Time: O(n^2), Omega(n)
	Space: O(1)

	A sequence can be sorted online by calling insertionSort
	multiple times while passing in the startIndex of newly added items.
	
	:param: sequence The sequence to sort in place.
	:param: startIndex The index to start sorting at.
 */
func insertionSort(sequence: inout [Int], startIndex: Int=1) {
	for advancingIndex in startIndex..<sequence.count {
		let key = sequence[advancingIndex]
		var retreatingIndex = advancingIndex - 1

		while retreatingIndex >= 0 && sequence[retreatingIndex] > key {
			sequence[retreatingIndex + 1] = sequence[retreatingIndex]
			retreatingIndex -= 1
		}

		sequence[retreatingIndex + 1] = key
	}
}

File: pythonPrep/sources/sorting/insertion_sort.py

def insertion_sort(sequence):
	""" Sort the sequence in place. Time: Omega(n), O(n^2). Space: O(1) aux. """
	for advancing_index in range(len(sequence)):
		key = sequence[advancing_index]
		retreating_index = advancing_index - 1

		while retreating_index >= 0 and sequence[retreating_index] > key:
			sequence[retreating_index + 1] = sequence[retreating_index]
			retreating_index -= 1

		sequence[retreating_index + 1] = key
